astropy/utils/data_info.py

```diff
@@ -0,0 +1,303 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# Note: these functions and classes are tested extensively in astropy table
+# tests via their use in providing mixin column info, and in
+# astropy/tests/test_info for providing table and column info summary data.
+
+import os
+import sys
+from copy import deepcopy
+import weakref
+import numpy as np
+from functools import partial
+
+from ..extern import six
+from ..utils import OrderedDict
+from ..utils.compat import NUMPY_LT_1_8
+
+def data_info_factory(names, funcs):
+    """
+    Factory to create a function that can be used as an ``option``
+    for outputting data object summary information.
+
+    Examples
+    --------
+    >>> from astropy.utils.data_info import data_info_factory
+    >>> from astropy.table import Column
+    >>> c = Column([4., 3., 2., 1.])
+    >>> mystats = data_info_factory(names=['min', 'median', 'max'],
+    ...                             funcs=[np.min, np.median, np.max])
+    >>> c.info(option=mystats)
+    min = 1.0
+    median = 2.5
+    max = 4.0
+    n_bad = 0
+    length = 4
+
+    Parameters
+    ----------
+    names: list
+        List of information attribute names
+    funcs: list
+        List of functions that compute the corresponding information attribute
+
+    Returns
+    -------
+    func: function
+        Function that can be used as a data info option
+    """
+    def func(dat):
+        outs = []
+        for name, func in zip(names, funcs):
+            try:
+                if isinstance(func, six.string_types):
+                    out = getattr(dat, func)()
+                else:
+                    out = func(dat)
+            except:
+                outs.append('--')
+            else:
+                outs.append(str(out))
+
+        return OrderedDict(zip(names, outs))
+    return func
+
+
+def _get_data_attribute(dat, attr=None):
+    """
+    Get a data object attribute for the ``attributes`` info summary method
+    """
+    if attr == 'class':
+        val = type(dat).__name__
+    elif attr == 'dtype':
+        val = dat.info.dtype.name
+    elif attr == 'shape':
+        datshape = dat.shape[1:]
+        val = datshape if datshape else ''
+    else:
+        val = getattr(dat.info, attr)
+    if val is None:
+        val = ''
+    return str(val)
+
+
+class InfoDescriptor(object):
```

Comment:
What I mean is that with the bound/unbound descriptor pattern, if you want to be able to something in the class's `__init__`, that `__init__` needs to handle being called in two different contexts:

1) class definition, in which case you're initializing an unbound instance of the descriptor--this is the one that lives on the class itself and provides the descriptor protocol:
```python
class MyClass:
    info = DataInfo()  # <-- DataInfo.__init__ called here
```
2) Creating a bound instance--when `__get__` is called via the descriptor protocol, the idea is that you then create a *new* `DataInfo` instance that is bound to some instance of the parent class.  So if you want the `DataInfo.__init__` to be able to do something different for bound instances you have to check in the `__init__` for if `_parent_ref` is set or not.

An alternative is to use different subclasses for the Unbound and Bound versions.  That might be simpler to see.

See how the [`astropy.modeling.Parameter`](http://docs.astropy.org/en/stable/api/astropy.modeling.Parameter.html#astropy.modeling.Parameter) class is implemented for reference.  I think ultimately I'm trying to point out that since we already use this pattern elsewhere in Astropy it would make sense to use generally where appropriate.  Maybe I can figure out a way to make it easy to implement this pattern generically, like I did with [`OrderedDescriptor`](http://docs.astropy.org/en/latest/api/astropy.utils.misc.OrderedDescriptor.html#astropy.utils.misc.OrderedDescriptor)